drivers/UARTSerial.cpp

```diff
@@ -347,6 +345,69 @@ void UARTSerial::tx_irq(void)
     }
 }
 
+/* These are all called from critical section */
+void UARTSerial::enable_rx_irq()
+{
+    SerialBase::attach(callback(this, &UARTSerial::rx_irq), RxIrq);
+    _rx_irq_enabled = true;
+}
+
+void UARTSerial::disable_rx_irq()
+{
+    SerialBase::attach(NULL, RxIrq);
+    _rx_irq_enabled = false;
+}
+
+void UARTSerial::enable_tx_irq()
+{
+    SerialBase::attach(callback(this, &UARTSerial::tx_irq), TxIrq);
+    _tx_irq_enabled = true;
+}
+
+void UARTSerial::disable_tx_irq()
+{
+    SerialBase::attach(NULL, TxIrq);
+    _tx_irq_enabled = false;
+}
+
+int UARTSerial::enable_input(bool yes)
+{
+    core_util_critical_section_enter();
+    if (_rx_enabled != yes) {
+        if (yes) {
+            UARTSerial::rx_irq();
+            if (!_rxbuf.full()) {
+                enable_rx_irq();
+            }
+        } else {
+            disable_rx_irq();
+        }
+        _rx_enabled = yes;
+    }
+    core_util_critical_section_exit();
+
+    return 0;
+}
+
+int UARTSerial::enable_output(bool yes)
+{
+    core_util_critical_section_enter();
+    if (_tx_enabled != yes) {
+        if (yes) {
+            UARTSerial::tx_irq();
+            if (_txbuf.full()) {
```

Comment:
This should probably be `if (!_txbuf.empty())`?

Reply:
Good spot.

I thought I'd already fixed it, except I've just realised that went to the wrong place - my topic branch is on the master repo not the fork. Oops. Abusing maintainer powers.

I guess changing that now will just mess things up.